--- main/codec_builtin.c	(Asterisk 13.7.2)
+++ main/codec_builtin.c	(working copy)
@@ -33,2 +33,3 @@
 
+#include "asterisk/ilbc.h"
 #include "asterisk/logger.h"
@@ -586,5 +587,10 @@
 static int ilbc_samples(struct ast_frame *frame)
 {
-	return 240 * (frame->datalen / 50);
+	struct ilbc_attr *attr = ast_format_get_attribute_data(frame->subclass.format);
+	const unsigned int mode = attr ? attr->mode : 30;
+	const unsigned int samples_per_frame = mode * ast_format_get_sample_rate(frame->subclass.format) / 1000;
+	const unsigned int octets_per_frame = (mode == 20) ? 38 : 50;
+
+	return samples_per_frame * frame->datalen / octets_per_frame;
 }
 
@@ -595,8 +601,8 @@
 	.sample_rate = 8000,
-	.minimum_ms = 30,
-	.maximum_ms = 300,
-	.default_ms = 30,
-	.minimum_bytes = 50,
+	.minimum_ms = 20,
+	.maximum_ms = 300,
+	.default_ms = 20,
+	.minimum_bytes = 38,
 	.samples_count = ilbc_samples,
-	.smooth = 1,
+	.smooth = 0,
 };
